src/cluster/Backend.h

```diff
@@ -0,0 +1,457 @@
+
+    /**
+     * Unregister interest in messages on a certain topic.
+     *
+     * @param topic_prefix a prefix previously supplied to Subscribe()
+     * @return true if interest in topic prefix is no longer advertised.
+     */
+    bool Unsubscribe(const std::string& topic_prefix) { return DoUnsubscribe(topic_prefix); }
+
+    /**
+     * Publish multiple log writes.
+     *
+     * All log records belong to (the stream, filter, path) pair that is
+     * described by \a header.
+     *
+     * @param header fixed information about the stream, writer, filter and the
+     * schema.
+     * @param path Separate from the header. One header may log to multiple paths,
+     * but the header fields are constant.
+     * @param records A span of logging::detail::LogRecords
+     */
+    bool PublishLogWrites(const zeek::logging::detail::LogWriteHeader& header,
+                          zeek::Span<zeek::logging::detail::LogRecord> records) {
+        return DoPublishLogWrites(header, records);
+    }
+
+protected:
+    /**
+     * Constructor.
+     */
+    Backend(std::unique_ptr<EventSerializer> es, std::unique_ptr<LogSerializer> ls)
+        : event_serializer(std::move(es)), log_serializer(std::move(ls)) {}
+
+    /**
+     * Process an incoming event message.
+     */
+    bool ProcessEventMessage(const std::string_view& topic, const std::string_view& format,
+                             detail::byte_buffer_span payload);
+
+    /**
+     * Process an incoming log message.
+     */
+    bool ProcessLogMessage(const std::string_view& format, detail::byte_buffer_span payload);
+
+private:
+    /**
+     * Called after all Zeek scripts have been loaded.
+     *
+     * A cluster backend should initialize itself based on script variables,
+     * register any IO sources. It should not yet start any connections, that
+     * should happen in DoInit() instead.
+     */
+    virtual void DoInitPostScript() = 0;
+
+    /**
+     * Called from Cluster::Backend::__init().
+     *
+     * Backend implementations should start connections with
+     * remote systems or other nodes, open listening ports or
+     * do whatever is needed to be functional.
+     */
+    virtual bool DoInit() = 0;
+
+    /**
+     * Called at termination time.
+     *
+     * This should be used to shut down connectivity. Any last messages
+     * to be published should be sent from script land, rather than in
+     * DoTerminate(). A backend may wait for a bounded and configurable
+     * amount of time to flush any last messages out.
+     */
+    virtual void DoTerminate() = 0;
+
+    /**
+     * Given the arguments for the \a Cluster::make_event() function, create
+     * a script-level record value that represents the event.
+     *
+     * The default implementation produces the \a Cluster::Event script type,
+     * but may be overridden by implementations.
+     *
+     * Note that this method exists primarily for the broker backend. Do not
```

Comment:
Related to Tim's question if that should eventually go away:

`Broker::make_event()` returns a `Broker::Event` and `Cluster::make_event()` will also return a `Broker::Event` when running with the Broker backend enabled. (Motivation here is that you can `s/Broker::make_event()/Cluster::make_event()/g` and not jeopardize Broker requiring an extra conversion.

> When would I need to reimplement it?

If `Cluster::Event` doesn't fit your purpose, but hopefully never. The motivation is a bit reversed, however: The new backends don't use `Broker::Event`.

If we switch broker over to using `Cluster::Event` and moving serialization out of broker, this can go away, but I didn't want to go there at this point.

Reply:
After pondering: I start to think this is confusing and complex and `Cluster::make_event()` should produce the same event for all backends.

My only worry is that something like the following (which today is working but seemingly barely used) would break at runtime when using a non-broker backend.
```
local e = Broker::make_event(...)
Cluster::publish_hrw(topic, key, e);
```

But maybe a `Reporter::error()` is enough for these cases. And even more maybe we could attempt to detect these via AST traversal.

Reply:
I have removed `MakeEvent()` and `DoMakeEvent()` from the `Backend` class and moved most logic into BifSupport.cc. This feels better even with the above being a bit quirky.

The fixups live in `topic/awelzel/pluggable-cluster-backends-part4-with-fixups`, but I'll fold them in soon.

Reply:
Cool, thanks a bunch. I get it now. And I like that iteration a lot. :+1: